src/include/Network/UPnP/ServiceClass.h

```diff
@@ -0,0 +1,70 @@
+/**
+ * ServiceClass.h
+ *
+ * Copyright 2020 mikee47 <mike@sillyhouse.net>
+ *
+ * This file is part of the Sming UPnP Library
+ *
+ * This library is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU General Public License as published by the Free Software Foundation, version 3 or later.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with Sming UPnP.
+ * If not, see <https://www.gnu.org/licenses/>.
+ *
+ ****/
+
+#pragma once
+
+#include "ClassObject.h"
+#include "Service.h"
+
+namespace UPnP
+{
+class DeviceControl;
+class ServiceControl;
+class DeviceClass;
+
+/**
+ * @brief Provides all information required for UPnP to construct a ServiceControl object
+ */
+class ServiceClass : public ClassObject
+{
+	friend DeviceControl;
+
+public:
+	using List = ObjectList<ServiceClass>;
+	using Field = Service::Field;
+
+	ServiceClass(const DeviceClass& deviceClass) : devcls(deviceClass)
+	{
+	}
+
+	Urn getServiceType() const
+	{
+		return ServiceUrn(getField(Field::domain), getField(Field::type), version());
+	}
+
+	virtual String getField(Field desc) const;
+
+	const ServiceClass* getNext() const
+	{
+		return reinterpret_cast<const ServiceClass*>(next());
+	}
+
+	const DeviceClass& deviceClass() const
+	{
+		return devcls;
+	}
+
+protected:
+	virtual ServiceControl* createObject(DeviceControl& device, const ServiceClass& serviceClass) const = 0;
```

Comment:
I guess you will need here also an empty virtual destructor. See https://gcc.gnu.org/faq.html#vtables.

Reply:
Are you getting compiler errors? ServiceClass inherits (eventually) from `Item` which has the virtual destructor, so should be sufficient?

Reply:
> Are you getting compiler errors? ServiceClass inherits

Yes. Multiple ```undefined reference to `vtable for` errors```. Compiler GCC 9. Compiling arguments:
`make SMING_ARCH=Host ENABLE_MALLOC_COUNT=0 ENABLE_GDB=1 DEBUG_VERBOSE_LEVEL=3`

GCC version:
`gcc version 9.3.0 (Ubuntu 9.3.0-10ubuntu2~16.04) `


```
Basic_ControlPoint: Linking out/Host/debug/firmware/app
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/samples/Basic_ControlPoint/out/Host/debug/lib/clib-App-d5f788d6d1c1dd58d53d4304b370719a.a(application.o): In function `_ZN4UPnP16schemas_upnp_org7service17RenderingControl15ClassCI2NS_12ServiceClassEERKNS_11DeviceClassE':
/home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/out/common/Network/UPnP/schemas-upnp-org/service/RenderingControl1.h:30: undefined reference to `vtable for UPnP::schemas_upnp_org::service::RenderingControl1::Class'
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/samples/Basic_ControlPoint/out/Host/debug/lib/clib-App-d5f788d6d1c1dd58d53d4304b370719a.a(application.o): In function `_ZN4UPnP16schemas_upnp_org7service18ConnectionManager15ClassCI2NS_12ServiceClassEERKNS_11DeviceClassE':
/home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/out/common/Network/UPnP/schemas-upnp-org/service/ConnectionManager1.h:30: undefined reference to `vtable for UPnP::schemas_upnp_org::service::ConnectionManager1::Class'
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/samples/Basic_ControlPoint/out/Host/debug/lib/clib-App-d5f788d6d1c1dd58d53d4304b370719a.a(application.o): In function `_ZN4UPnP16schemas_upnp_org7service12AVTransport15ClassCI2NS_12ServiceClassEERKNS_11DeviceClassE':
/home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/out/common/Network/UPnP/schemas-upnp-org/service/AVTransport1.h:30: undefined reference to `vtable for UPnP::schemas_upnp_org::service::AVTransport1::Class'
```

Reply:
It's probably the intermedia source. Try:

```
make UPnP-build
make
```

Reply:
Need to fix that compile issue, and consider how to handle generated source within the framework.